pkg/controller/machine-set-boot-image/helpers.go

```diff
@@ -0,0 +1,108 @@
+package machineset
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	archtranslater "github.com/coreos/stream-metadata-go/arch"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	opv1 "github.com/openshift/api/operator/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+)
+
+// TODO - unmarshal the providerspec into each ProviderSpec type until it succeeds,
+// and then call the appropriate reconcile function. This is needed for multi platform
+// support
+func unmarshalToFindPlatform(machineSet *machinev1beta1.MachineSet, _ *corev1.ConfigMap, arch string) (patchRequired bool, newMachineSet *machinev1beta1.MachineSet, err error) {
+	klog.Infof("Skipping machineset %s, unknown platform type with %s arch", machineSet.Name, arch)
+	return false, nil, nil
+}
+
+// This function unmarshals the machineset's provider spec into
+// a ProviderSpec object. Returns an error if providerSpec field is nil,
+// or the unmarshal fails
+func unmarshalProviderSpec(ms *machinev1beta1.MachineSet, providerSpec interface{}) error {
+	if ms.Spec.Template.Spec.ProviderSpec.Value == nil {
+		return fmt.Errorf("providerSpec field was empty")
+	}
+	if err := yaml.Unmarshal(ms.Spec.Template.Spec.ProviderSpec.Value.Raw, &providerSpec); err != nil {
+		return fmt.Errorf("unmarshal into providerSpec failedL %w", err)
+	}
+	return nil
+}
+
+// This function marshals the ProviderSpec object into a MachineSet object.
+// Returns an error if ProviderSpec or MachineSet is nil, or if the marshal fails
+func marshalProviderSpec(ms *machinev1beta1.MachineSet, providerSpec interface{}) error {
+	if ms == nil {
+		return fmt.Errorf("MachineSet object was nil")
+	}
+	if providerSpec == nil {
+		return fmt.Errorf("ProviderSpec object was nil")
+	}
+	rawBytes, err := json.Marshal(providerSpec)
+	if err != nil {
+		return fmt.Errorf("marshal into machineset failed: %w", err)
+	}
+	ms.Spec.Template.Spec.ProviderSpec.Value = &kruntime.RawExtension{Raw: rawBytes}
```

Comment:
nit: are we guaranteed that every field in the chain `ms.Spec.Template.Spec.ProviderSpec.Value` is not nil if `ms!=nil`? We can still panic if say, ProviderSpec doesn't exist, or is everything a required field?

Reply:
Hmm, I guess it is a hassle to do a chain'ed if check to see if everything under here exists, but I'm just concerned that e.g. ms.Spec.Template.Spec exists, but ms.Spec.Template.Spec.ProviderSpec is empty and we panic here, which is not idea.

Your call though :)